File: src/year2020/day20/day20.py
#!/usr/bin/env python3
# [Day 20: Jurassic Jigsaw](https://adventofcode.com/2020/day/20)

import sys
from collections import Counter
from copy import deepcopy
from functools import reduce
from operator import mul
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()

tiles = {}

for tile in data.split("\n\n"):
    tile = tile.splitlines()

    assert len(tile) == 1 + 10
    assert all(len(row) == 10 for row in tile)

    # "Tile 3011:"
    id = int(tile[0][5:-1])

    tile = [[c for c in row] for row in tile[1:]]

    t = []

    for _ in range(2):
        for _ in range(4):
            # read the binary signature of the tile (first row)
            b0 = 0
            for bit in range(10):
                # up
                if tile[0][bit] == "#":
                    b0 += 1 << bit
            t.append((b0, deepcopy(tile)))

            # rotate
            tile2 = [[0 for _ in range(10)] for _ in range(10)]
            for x in range(10):
                for y in range(10):
                    tile2[y][x] = tile[9 - x][y]
            tile = tile2

        # flip
        for y in range(10):
            for x in range(5):
                tile[y][x], tile[y][9 - x] = tile[y][9 - x], tile[y][x]

    # { id: [(sig, tile), (sig, tile), ...], }
    tiles[id] = t


# part 1

counts = list()
for k, v in tiles.items():
    for sig, _ in v:
        counts.append(sig)
counts = Counter(counts)
# counts[]==1 <=> tile on map boundaries. if count of 4, it's a corner
corners = list(id for id, v in tiles.items() if sum(1 for k, _ in v if counts[k] == 1) == 4)
print(reduce(mul, corners))


# part 2

N = int(len(tiles) ** 0.5)
big_grid = [["?" for _ in range(N * 10)] for _ in range(N * 10)]
monster_grid = [["?" for _ in range(N * 8)] for _ in range(N * 8)]


def reverse_sig(b):
    """Reverse the binary signature."""
    r = 0
    for i in range(10):
        r *= 2
        if b & 1 == 1:
            r += 1
        b //= 2
    return r


def opposite_edge(id, sig):
    """Get the opposite edge of a tile relative to a sig."""
    t = tiles[id]
    sig = reverse_sig(sig)
    for i, (b, _) in enumerate(t):
        if sig == b:
            return t[((i % 4) + 2) % 4 + 4 * (i // 4)][0]


def adj_edge(id, sig):
    """Get the left edge of a tile relative to a sig."""
    t = tiles[id]

    for i, (b, _) in enumerate(t):
        if sig == b:
            return t[((i % 4) + 3) % 4 + 4 * (i // 4)][0]


def print_tile(id, top_sig, x, y):
    """Print the tile into the map."""
    # print(id, top_sig, bin(top_sig))
    for b, v in tiles[id]:
        if top_sig == b:
            # print("\n".join("".join(c for c in row) for row in v))

            for i in range(10):
                for j in range(10):
                    big_grid[i + y * 10][j + x * 10] = v[i][j]

            for i in range(8):
                for j in range(8):
                    monster_grid[i + y * 8][j + x * 8] = v[i + 1][j + 1]

            break


def next_tile(prev_id, sig):
    """Find the next tile (same sig, but different id)."""
    for id, v in tiles.items():
        if id != prev_id:
            for b, _ in v:
                if b == sig:
                    return id


edges = []
id = corners[0]
for i in range(4):
    sig1 = tiles[id][i][0]
    sig2 = tiles[id][(i + 1) % 4][0]
    if counts[sig1] == 2 and counts[sig2] == 2:
        edges.append(sig1)
        edges.append(sig2)
        break


x, y = 0, 0

id_top = id
bottom = opposite_edge(id, edges[1])
right = edges[0]


# fill the map
while True:
    y = 0
    while id:
        print_tile(id, bottom, x, y)
        y += 1
        bottom = opposite_edge(id, bottom)
        id = next_tile(id, bottom)

    x += 1
    id_top = next_tile(id_top, right)
    if not id_top:
        break
    id = id_top
    bottom = adj_edge(id, right)
    right = opposite_edge(id, right)


if verbose:
    for y, row in enumerate(big_grid):
        s = ""
        for x, c in enumerate(row):
            s += c
            if x % 10 == 9:
                s += " "
        print(s)
        if y % 10 == 9:
            print()


monster_def = """\
                  #
#    ##    ##    ###
 #  #  #  #  #  #
"""

# parse the monster
monster_def = monster_def.splitlines()
MX = max(len(line) for line in monster_def)
MY = len(monster_def)
monster = []
monster_length = 0
for y in range(MY):
    m = []
    for x in range(MX):
        if x < len(monster_def[y]):
            if monster_def[y][x] == "#":
                monster_length += 1
            m.append(monster_def[y][x])
        else:
            m.append(" ")
    monster.append(m)

# look for the monster
found = 0
for _ in range(2):
    for _ in range(4):
        # search
        for y in range(N * 8 - MY):
            for x in range(N * 8 - MX):
                m = 0
                for my in range(MY):
                    for mx in range(MX):
                        if monster_grid[y + my][x + mx] == monster[my][mx] == "#":
                            m += 1
                if monster_length == m:
                    found += 1

        # rotate
        rotate = [[0 for _ in range(N * 8)] for _ in range(N * 8)]
        for x in range(N * 8):
            for y in range(N * 8):
                rotate[y][x] = monster_grid[N * 8 - 1 - x][y]
        monster_grid = rotate

    # flip
    for y in range(N * 8):
        for x in range(N * 8 // 2):
            monster_grid[y][x], monster_grid[y][N * 8 - 1 - x] = monster_grid[y][N * 8 - 1 - x], monster_grid[y][x]

# count the '#' on the map
hashtag = 0
for y in range(N * 8):
    for x in range(N * 8):
        if monster_grid[y][x] == "#":
            hashtag += 1

# at last...
print(hashtag - found * monster_length)

File: 2020/day20/day20.rs
//! [Day 20: Jurassic Jigsaw](https://adventofcode.com/2020/day/20)

use aoc::math::UnsignedMathOps;
use aoc::{Counter, GridU, Square};
use rustc_hash::FxHashMap;

const NESSIE: &str = concat!(
    "                  # \n",
    "#    ##    ##    ###\n",
    " #  #  #  #  #  #   \n"
);

/// The tile signature is the binary value of its top row. It has to be unique (and hopefully it is).
/// Since the tils are 10 pixel wide, a u16 is sufficient.
type TileSig = u16;

type Tile = Square<u8>;

struct TileArrangement(Vec<(TileSig, Tile)>);

fn calc_sig(p: &Tile) -> TileSig {
    let first_row = p.iter_rows().next().unwrap();

    first_row
        .iter()
        .fold(0, |acc, pixel| (acc << 1) | TileSig::from(pixel == &b'#'))
}

/// Reverse the binary signature.
fn symmetric_sig(mut b: TileSig) -> TileSig {
    let mut r = 0;
    for _ in 0..10 {
        r *= 2;
        if b & 1 == 1 {
            r += 1;
        }
        b /= 2;
    }
    r
}

struct LochNess {
    monster_map: Square<u8>,
}

impl LochNess {
    fn new(n: usize) -> Self {
        Self {
            monster_map: Square::new(n * 8),
        }
    }

    /// Reconstruct the entire image with the fixed image tile
    /// at the its position
    fn add_fixed_tile(&mut self, tile: &TileArrangement, sig: TileSig, x: usize, y: usize) {
        //
        for (s, v) in &tile.0 {
            if *s == sig {
                for i in 0..8 {
                    for j in 0..8 {
                        self.monster_map[(x * 8 + i, y * 8 + j)] = v[(i + 1, j + 1)];
                    }
                }
                return;
            }
        }
        panic!()
    }

    /// Return the occurence of monster in the image.
    fn find_monster(&self, monster: &GridU<u8>) -> usize {
        let n = self.monster_map.size();
        let (width, height) = monster.size();

        let mut count = 0;

        for my in 0..(n - height) {
            for mx in 0..(n - width) {
                count += usize::from(
                    monster
                        .iter()
                        .all(|((x, y), &m)| m != b'#' || m == self.monster_map[(mx + x, my + y)]),
                );
            }
        }

        count
    }
}

struct Puzzle {
    tiles: FxHashMap<u32, TileArrangement>,
}

impl Puzzle {
    /// Find the `tile_id` for the four corners of the map.
    fn corners(&self) -> (Vec<u32>, Counter<TileSig>) {
        // counts the signature of each tile in any position
        // counts[]==1 <=> tile is on map boundaries.
        // if count of 4, it's a corner

        let mut counts = Counter::<_>::new();
        for tile in self.tiles.values() {
            for (sig, _) in &tile.0 {
                counts.insert(*sig);
            }
        }

        let corners = self
            .tiles
            .iter()
            .filter_map(|(&tile_id, tile)| {
                let sig_count = tile
                    .0
                    .iter()
                    .filter_map(|(sig, _)| (counts[sig] == 1).then_some(1))
                    .count();

                (sig_count == 4).then_some(tile_id)
            })
            .collect();

        (corners, counts)
    }

    /// Get the opposite edge of a tile relative to a sig.
    fn opposite_edge(&self, id: u32, sig: TileSig) -> TileSig {
        let t = &self.tiles[&id];
        let sig = symmetric_sig(sig);
        for (i, (b, _)) in t.0.iter().enumerate() {
            if sig == *b {
                let opposite_idx = ((i % 4) + 2) % 4 + 4 * (i / 4);
                return t.0[opposite_idx].0;
            }
        }
        panic!()
    }

    /// Get the left edge of a tile relative to a sig.
    fn adjacent_edge(&self, id: u32, sig: TileSig) -> TileSig {
        let t = &self.tiles[&id];
        for (i, (b, _)) in t.0.iter().enumerate() {
            if sig == *b {
                let adj_idx = ((i % 4) + 3) % 4 + 4 * (i / 4);
                return t.0[adj_idx].0;
            }
        }
        panic!()
    }

    /// Find the next tile (same sig, but different id).
    fn next_tile(&self, prev_id: u32, sig: TileSig) -> Option<u32> {
        for (id, v) in &self.tiles {
            if id != &prev_id {
                for (square_sig, _) in &v.0 {
                    if square_sig == &sig {
                        return Some(*id);
                    }
                }
            }
        }
        None
    }
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        let mut tiles = FxHashMap::default();

        for tile_def in data.trim_ascii().split("\n\n") {
            let (tile_id, grid) = tile_def.split_once('\n').unwrap();

            // "Tile XXX:"
            let tile_id = tile_id
                .strip_prefix("Tile ")
                .unwrap()
                .trim_end_matches(':')
                .parse()
                .unwrap();

            let square = Square::parse(grid, '\n');

            let mut tile = TileArrangement(Vec::with_capacity(8));

            for p in square.iter_pos() {
                let sig = calc_sig(&p);
                tile.0.push((sig, p));
            }

            tiles.insert(tile_id, tile);
        }

        Self { tiles }
    }

    /// Solve part one.
    fn part1(&self) -> u64 {
        self.corners().0.iter().copied().map(u64::from).product()
    }

    /// Solve part two.
    fn part2(&self) -> usize {
        let n = self.tiles.len().sqrt();
        let mut lochness = LochNess::new(n);

        // find the first corner
        // the both clockwise sigs have be count of 2
        let mut edge1 = 0;
        let mut edg2 = 0;
        let (corners, counts) = self.corners();
        let mut id_top = corners[0];
        let corner = &self.tiles[&id_top];

        for i in 0..4 {
            let sig1 = corner.0[i].0;
            let sig2 = corner.0[(i + 1) % 4].0;

            if counts[&sig1] == 2 && counts[&sig2] == 2 {
                edge1 = sig1;
                edg2 = sig2;
                break;
            }
        }

        // reconstruct the map

        let mut bottom = self.opposite_edge(id_top, edg2);
        let mut right = edge1;
        let mut x = 0;
        loop {
            let mut id = id_top;
            let mut y = 0;

            loop {
                lochness.add_fixed_tile(&self.tiles[&id], bottom, x, y);
                y += 1;

                bottom = self.opposite_edge(id, bottom);
                if let Some(next_id) = self.next_tile(id, bottom) {
                    id = next_id;
                } else {
                    break;
                }
            }

            x += 1;
            if let Some(next_id_top) = self.next_tile(id_top, right) {
                id_top = next_id_top;
                id = id_top;
                bottom = self.adjacent_edge(id, right);
                right = self.opposite_edge(id, right);
            } else {
                break;
            }
        }

        // look for the monster
        let nessie = GridU::<u8>::parse(NESSIE);

        let found: usize = nessie.iter_pos().map(|r| lochness.find_monster(&r)).sum();

        let rough = lochness
            .monster_map
            .iter()
            .filter(|(_, &c)| c == b'#')
            .count();

        let monster_length = nessie.iter().filter(|(_, &c)| c == b'#').count();

        rough - found * monster_length
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (u64, usize) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn part1() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 20899048083289);
    }

    #[test]
    fn part2() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 273);
    }
}
